crates/re_viewer/src/ui/view_spatial/ui_2d.rs

```diff
@@ -247,198 +269,231 @@ pub fn view_2d(
     auto_size_config: re_renderer::AutoSizeConfig,
     any_outlines: bool,
+    pinhole: Option<Pinhole>,
 ) -> anyhow::Result<TargetConfiguration> {
     let pixels_from_points = painter.ctx().pixels_per_point();
     let resolution_in_pixel =
         gpu_bridge::viewport_resolution_in_pixels(painter.clip_rect(), pixels_from_points);
     anyhow::ensure!(resolution_in_pixel[0] > 0 && resolution_in_pixel[1] > 0);
 
-    let camera_position_space = space_from_ui.transform_pos(painter.clip_rect().min);
+    // TODO(#1988):
+    // The camera setup is done in a way that works well with the way we inverse pinhole camera transformations right now.
+    // This has a lot of issues though, mainly because we pretend that the 2D plane has a defined depth.
+    // * very bad depth precision as we limit the depth range from 0 to focal_length_in_pixels
+    // * depth values in depth buffer are almost non-sensical and can't be used easily for picking
+    // * 2D rendering can use depth buffer for layering only in a very limited way
+    //
+    // Instead we should treat 2D objects as pre-projected with their depth information already lost.
+    //
+    // We would define two cameras then:
+    // * an orthographic camera for handling 2D rendering
+    // * a perspective camera *at the origin* for 3D rendering
+    // Both share the same view-builder and the same viewport transformation but are independent otherwise.
+
+    // For simplicity (and to reduce surprises!) we always render with a pinhole camera.
+    // Make up a default pinhole camera if we don't have one placing it in a way to look at the entire space.
+    let canvas_size = glam::vec2(canvas_from_ui.to().width(), canvas_from_ui.to().height());
+    let default_principal_point = canvas_size * 0.5;
+    let pinhole = pinhole.unwrap_or_else(|| {
+        let focal_length_in_pixels = canvas_size.x;
+
+        re_log_types::Pinhole {
+            image_from_cam: glam::Mat3::from_cols(
+                glam::vec3(focal_length_in_pixels, 0.0, 0.0),
+                glam::vec3(0.0, focal_length_in_pixels, 0.0),
+                default_principal_point.extend(1.0),
+            )
+            .into(),
+            resolution: Some(canvas_size.into()),
+        }
+    });
+
+    let projection_from_view = re_renderer::view_builder::Projection::Perspective {
+        vertical_fov: pinhole.fov_y().unwrap_or(Eye::DEFAULT_FOV_Y),
+        near_plane_distance: 0.01,
+        aspect_ratio: pinhole
+            .aspect_ratio()
+            .unwrap_or(canvas_size.x / canvas_size.y),
+    };
+
+    // Put the camera at the position where it sees the entire image plane as defined
+    // by the pinhole camera.
+    let view_from_world = macaw::IsoTransform::look_at_rh(
+        pinhole
+            .principal_point()
+            .extend(-pinhole.focal_length_in_pixels()),
+        pinhole.principal_point().extend(0.0),
+        -glam::Vec3::Y,
+    )
+    .unwrap_or(macaw::IsoTransform::IDENTITY);
```

Comment:
I think the fallback here should be to log an error and return, not use a rather arbitrary transform